quil-rs/src/parser/lexer/mod.rs

```diff
@@ -135,6 +135,13 @@ pub(crate) fn lex(input: LexInput) -> Result<Vec<TokenWithLocation>, LexError> {
         .map_err(LexError::from)
 }
 
+/// TODO: Discuss:
+///     * Should we make indentation exactly 4 spaces, rather than 4-7?
+///     * Should we allow for multiple levels of indentation? The Quil spec implies that only a
+///       single level indentation is possible, as each indentation must be preceded by a newline
+///     * What are the implications of using spaces instead of tabs for program serialization?
+///       It is technically 4x the bytes for indentations in serialized program. Seems minor, but
+///       maybe tabs were chosen for a reason?
```

Comment:
I am in favor of using exactly 4 spaces for indentation, as that most closely aligns with the Quil spec.

Not allowing for multiple levels of indentation would also be in line with the Quil spec, but Program serialization currently allows for that (e.g by adding a DEFCIRCUIT to another DEFCIRCUIT), these cases would fail to roundtrip. That is a potential UX issue, but could probably be solved by checking for nested indentation in any instruction that indents a body of instructions.

Reply:
1. I am also in favor of enforcing exactly 4 spaces
2. I'd expect nested indentation not to be used in practice anywhere, so I don't _think_ disallowing nested indentation would break much. That said, we _do_ recursively expand calibrations, so it's possible people are using nested definitions and having it work. I think it's worth maybe discussing further but not making a change here.
3. I tested this locally using a large program from [quilc](https://github.com/quil-lang/quilc) and a Rigetti calibrations file:

```
== quilc/benchmarking/quil-available/0008q-0019430-grover8bit.quil ==
Tabs:         4916568
Tabs (gz):      38937
Spaces:       4946520
Spaces (gz):    39868
Size increase:  0.061%
Size inc. (gz): 0.23%

== Ankaa-2 Calibrations ==
Tabs:          460487
Tabs (gz):      50748
Spaces:        480647
Spaces (gz):    51077
Size increase:  4.38%
Size inc. (gz): 0.65%
```

I'm not sure that 4% is worth worrying about and, if it is, IMO we should rely on compression instead of tabs vs spaces.

Reply:
Per these discussions, and some offline, I've decided to use 4 spaces for serialization, but continue to support tabs in parsing to avoid breaking downstream users more than necessary. I also created [an issue](https://github.com/rigetti/quil-rs/issues/391) to handle multiple level of indentations.